hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/tools/TestDFSZKFailoverController.java

```diff
@@ -301,6 +304,40 @@ public void testManualFailoverWithDFSHAAdmin() throws Exception {
     waitForHAState(1, HAServiceState.STANDBY);
   }
 
+  /**
+   * Tests that a Namenode in safe mode should not be transfer to observer state.
+   */
+  @Test
+  public void testManualFailoverWithDFSHAAdminInSafemode() throws Exception {
+    startCluster();
+    NamenodeProtocols nn1 = cluster.getNameNode(1).getRpcServer();
+
+    // Enter safe mode.
+    nn1.setSafeMode(HdfsConstants.SafeModeAction.SAFEMODE_ENTER, false);
+    // Test NameNodeRpcServer.
+    LambdaTestUtils.intercept(SafeModeException.class,
+        "Cannot transition to observer. Name node is in safe mode",
+        () -> nn1.transitionToObserver(
+            new StateChangeRequestInfo(RequestSource.REQUEST_BY_USER_FORCED)));
```

Comment:
This should probably be in `TestHASafeMode`, where we already have `testTransitionToActiveWhenSafeMode`